bin/convert2vulyk.py

```diff
@@ -67,43 +97,156 @@ def parse_bsf(bsf_data: str) -> list:
     #                     Token_id Entity start  end    text within range
     #                        \/      \/     \/    \/      \/
     ln_ptrn = re.compile(r"(T\d+)\s(\w+)\s(\d+)\s(\d+)\s(.+?)(?=T\d+\s\w+\s\d+\s\d+|$)", flags=re.DOTALL)
-    result = []
+    result: list = []
     for m in ln_ptrn.finditer(data):
         bsf = BsfInfo(m.group(1), m.group(2), int(m.group(3)), int(m.group(4)), m.group(5).strip())
         result.append(bsf)
     return result
 
 
-def convert_bsf_2_vulyk(text: str, bsf_markup: str) -> dict:
+def simple_tokenizer(text: str) -> list[list[str]]:
+    """
+    Given whitespace/newline tokenized text, return the
+    list of sentences (where each sentence is made of tokens)
+    using whitespaces and newlines
+    """
+
+    doc: list[list[str]] = []
+    if text:
+        for s in text.split("\n"):
+            doc.append(s.split(" "))
+
+    return doc
+
+
+def read_and_tokenize(text: str, fmt: str, tokenizer: TokenizationType) -> list[list[str]]:
+    if fmt == "json":
+        assert tokenizer == TokenizationType.NOOP, "Json is meant to be already tokenized"
+        return json.loads(text)
+    else:
+        assert tokenizer != TokenizationType.NOOP, "You cannot keep texts not-tokenized"
+        if tokenizer == TokenizationType.WHITESPACE:
+            return simple_tokenizer(text)
+        elif tokenizer == TokenizationType.TOKENIZE_UK:
+            return list(chain(*tokenize_text(text)))
+
+    return []  # Calm down, mypy
+
+
+def reconstruct_tokenized(tokenized_text: list[list[str]]) -> Generator[AlignedToken, None, None]:
```

Comment:
Do I get it right that the only purpose of this function is to fix an issue of tokenizer when punctuation is improperly separated with space? Or is there something else that I miss?

Reply:
Yes. I'm a bit picky about that (since I was doing a lot of FE back in a days). That rabbit hole was much deeper than I thought (you can see the number of tests I wrote) but I'm glad I'm finally did it (however the time spent/result ratio was way below standards)